libsplinter/src/store/mod.rs

```diff
@@ -0,0 +1,95 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#[cfg(feature = "postgres")]
+pub mod postgres;
+#[cfg(feature = "sqlite")]
+pub mod sqlite;
+
+#[cfg(feature = "diesel")]
+use diesel::r2d2::{ConnectionManager, Pool};
+
+/// An abstract factory for creating Splinter stores backed by the same storage.
+pub trait StoreFactory {
+    /// Get a new `CredentialsStore`
+    #[cfg(feature = "biome-credentials")]
+    fn get_biome_credentials_store(&self) -> Box<dyn crate::biome::CredentialsStore>;
+
+    /// Get a new `KeyStore`
+    #[cfg(feature = "biome-key-management")]
+    fn get_biome_key_store(&self) -> Box<dyn crate::biome::KeyStore>;
+
+    /// Get a new `RefreshTokenStore`
+    #[cfg(feature = "biome-credentials")]
+    fn get_biome_refresh_token_store(&self) -> Box<dyn crate::biome::RefreshTokenStore>;
+
+    /// Get a new `UserStore`
+    fn get_biome_user_store(&self) -> Box<dyn crate::biome::UserStore>;
+}
+
+/// Creates a `StoreFactory` backed by the given storage location.
+///
+/// # Arguments
+///
+/// * `storage_location` - The type and location of the storage that will back all stores created
+///   by the resulting factory.
+pub fn create_store_factory(
+    storage_location: StorageLocation,
+) -> Result<Box<dyn StoreFactory>, StoreFactoryCreationError> {
+    match storage_location {
+        #[cfg(feature = "postgres")]
+        StorageLocation::Postgres(url) => {
+            let connection_manager = ConnectionManager::<diesel::pg::PgConnection>::new(url);
+            let pool = Pool::builder().build(connection_manager).map_err(|err| {
+                StoreFactoryCreationError(format!("Failed to build connection pool: {}", err))
+            })?;
+            Ok(Box::new(postgres::PgStoreFactory::new(pool)))
+        }
+        #[cfg(feature = "sqlite")]
+        StorageLocation::Sqlite(conn_str) => {
+            let connection_manager =
+                ConnectionManager::<diesel::sqlite::SqliteConnection>::new(&conn_str);
+            let mut pool_builder = Pool::builder();
+            // A new database is created for each connection to the in-memory SQLite
+            // implementation; to ensure that the resulting stores will operate on the same
+            // database, only one connection is allowed.
+            if conn_str == ":memory:" {
+                pool_builder = pool_builder.max_size(1);
+            }
+            let pool = pool_builder.build(connection_manager).map_err(|err| {
+                StoreFactoryCreationError(format!("Failed to build connection pool: {}", err))
+            })?;
+            Ok(Box::new(sqlite::SqliteStoreFactory::new(pool)))
+        }
+    }
+}
+
+pub enum StorageLocation {
```

Comment:
Don't like this enum name; however, I think it would be better to get rid of this enum completely and just match on the appropriate substring in the url.

Reply:
String matching causes a lot of complexity here. Which storage types are supported is determined by compilation flags; string matching forces us to check if the storage type is supported and handle any errors that may arise with that usage. With an enum, the compiler removes any possibility for errors; you can only use the enum variants that match your compilation options.

This is made even more complicated by SQLite supporting a variety of values (file paths, URIs, the `:memory:` string). It also assumes that every DB we implement support for can be identified by just the string passed in (which may not always be the case).

Reply:
We still need to design the CLI arguments (for daemon and CLIs (at least migration CLI)) -- I think we should reduce them as much as possible (if we have to specify a URL and the database type every time, we've failed). That will drive what we need here, though we could have a String-to-URL-enum function to partition the "unknown database" error a bit.

For now, I'd at least like to see the enum with a better name. How about Url or DatabaseUrl or DatabaseUri.

Reply:
What happens when you compile this with just store-factory?  Should there be a Memory option that's always available?

Reply:
Updated enum name to `ConnectionUri`

Reply:
@peterschwarz added the memory store factory (and variant of the enum)